### PR TITLE
cmake: use ninja job pool

### PR DESCRIPTION
**Pull request by tchaikov**
Signed-off-by: Kefu Chai <kchai@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
but we still miss the jobserver in ninja, see https://github.com/dockbuild/ninja-jobserver. probably, we could use use fork of ninja, see https://github.com/Kitware/ninja/releases

**Comment by tchaikov** [comment]
jenkins test classic perf

**Comment by tchaikov** [comment]
@smithfarm i don't expect this single change to address the OOM in the OBS builder. but i think, by introducing the job pool machinery, we can limit the parallelism level with a finer grained control based on the per-target basis.

**Comment by smithfarm** [comment]
> @smithfarm i don't expect this single change to address the OOM in the OBS builder. but i think, by introducing the job pool machinery, we can limit the parallelism level with a finer grained control on the per-target basis.

The problem I see is that Ninja itself seems to use more memory than make. So when we switch the RPM build from make to Ninja, the build becomes more memory-hungry, regardless of the parallelism setting.

**Comment by smithfarm** [comment]
(The OOM in OBS only happens with Ninja, not with make)

**Comment by tchaikov** [comment]
i see. another finding when looking into this issue is that, ninja does not have jobserver support yet. the child instance of ninja job is not managed by the parent ninja instance, so, for instance, when we build rocksdb external project using the ninja generator, it is built at **full speed**.

but we don't suffer from this, as GNU make is able to take care of this just fine with its job server.

**Comment by smithfarm** [comment]
Politically speaking, folks who build Ceph a lot can be divided into two groups, according to available memory on their build machines :-) What is good for one group is not necessarily good for the other.

Ceph seems very memory-intensive to build, and I'm not aware of any open-source project that is more memory-intensive to build. If Ceph really is the most memory-hungry project (or one of the most), would it make sense for Ceph to proceed with care when changing its build system and to maintain an awareness of how proposed changes will affect memory usage patterns? Such an awareness could inform judgments on the relative costs and benefits of such changes.

**Comment by tchaikov** [comment]
FWIW, i had to reboot my own building host last week-end at least for 4 times. for what reason? lacking patience and lacking memory =)

**Comment by tchaikov** [comment]
@smithfarm again, i hear you. and i am not ignoring the voices from community. actually, i am in one of group who don't have abundant memory. so i am trying to address this problem from time to time.

**Comment by smithfarm** [comment]
@tchaikov Thanks!

File: src/test/librbd/CMakeLists.txt [hunk 1]
```diff
@@ -158,6 +158,9 @@ add_executable(unittest_librbd
   ${unittest_librbd_srcs}
   $<TARGET_OBJECTS:common_texttable_obj>)
 target_compile_definitions(unittest_librbd PRIVATE "TEST_LIBRBD_INTERNALS")
+set_target_properties(unittest_librbd PROPERTIES
```
Comment (trociny) [inline src/test/librbd/CMakeLists.txt:161]: @tchaikov I expect unittest_rbd_mirror may also be a problem.

Comment (tchaikov) [reply src/test/librbd/CMakeLists.txt:161]: @trociny thanks for your inputs. i agree, we need to tackle this problem using a case-by-case approach. 

but on my building host, the source files included by `rbd_mirror_internal` target take around 800MiB memory on average. in my case:

```
    PID USER      PRI  NI  VIRT   RES   SHR S CPU%-MEM%   TIME+  Command
 657686 kefu       20   0  770M  744M 21612 R 100.  4.7  0:17.44 /usr/lib/gcc/x86_64-linux-gnu/11/cc1plus -quiet -I src/include -I ../src -imultiarch x86_64-linux-gnu -MD src/tools/rbd_mirror/CMakeFiles/rbd_mirror_internal.dir/image_replayer/snapshot/Utils.cc.d -MF src/to
 657869 kefu       20   0  801M  775M 20856 R 100.  4.9  0:14.76 /usr/lib/gcc/x86_64-linux-gnu/11/cc1plus -quiet -I src/include -I ../src -imultiarch x86_64-linux-gnu -MD src/tools/rbd_mirror/CMakeFiles/rbd_mirror_internal.dir/image_sync/SyncPointCreateRequest.cc.d -MF sr
 658177 kefu       20   0  608M  583M 20364 R 100.  3.7  0:09.91 /usr/lib/gcc/x86_64-linux-gnu/11/cc1plus -quiet -I src/include -I ../src -imultiarch x86_64-linux-gnu -MD src/tools/rbd_mirror/CMakeFiles/rbd_mirror_internal.dir/image_sync/SyncPointPruneRequest.cc.d -MF src
 658552 kefu       20   0  347M  318M 16124 R 99.5  2.0  0:03.43 /usr/lib/gcc/x86_64-linux-gnu/11/cc1plus -quiet -I src/include -I ../src -imultiarch x86_64-linux-gnu -MD src/tools/rbd_mirror/CMakeFiles/rbd_mirror_internal.dir/pool_watcher/RefreshImagesRequest.cc.d -MF sr
```

and the ones in unittest_rbd_mirror could take up to 1.9 GiB.

```
    PID USER      PRI  NI  VIRT   RES   SHR S CPU%-MEM%   TIME+  Command
 670667 kefu       20   0 1905M 1880M 24184 R 100. 11.8  0:59.42 /usr/lib/gcc/x86_64-linux-gnu/11/cc1plus -quiet -I src/include -I ../src -imultiarch x86_64-linux-gnu -MD src/test/rbd_mirror/CMakeFiles/unittest_rbd_mirror.dir/image_deleter/test_mock_SnapshotPurgeRequest.c
 671577 kefu       20   0 1550M 1522M 21080 R 100.  9.5  0:44.40 /usr/lib/gcc/x86_64-linux-gnu/11/cc1plus -quiet -I src/include -I ../src -imultiarch x86_64-linux-gnu -MD src/test/rbd_mirror/CMakeFiles/unittest_rbd_mirror.dir/image_deleter/test_mock_TrashMoveRequest.cc.d
 672319 kefu       20   0 1399M 1370M 21136 R 100.  8.6  0:33.00 /usr/lib/gcc/x86_64-linux-gnu/11/cc1plus -quiet -I src/include -I ../src -imultiarch x86_64-linux-gnu -MD src/test/rbd_mirror/CMakeFiles/unittest_rbd_mirror.dir/image_deleter/test_mock_TrashWatcher.cc.d -MF
 672136 kefu       20   0 1312M 1280M 21064 R 99.6  8.0  0:35.40 /usr/lib/gcc/x86_64-linux-gnu/11/cc1plus -quiet -I src/include -I ../src -imultiarch x86_64-linux-gnu -MD src/test/rbd_mirror/CMakeFiles/unittest_rbd_mirror.dir/image_deleter/test_mock_TrashRemoveRequest.cc.`
```

so, i am able to use `-j6`on my machine with 16 GB memory without worrying about OOM. for your reference, i am using GCC-11. i understand, your mileage may vary, so if you spot a very high memory usage when compiling `unittest_rbd_mirror`, i will be happy to add add it to the "heavy job" club.

Comment (trociny) [reply src/test/librbd/CMakeLists.txt:161]: No. It is just I know that rbd mirror unit tests are heavily using templates for mock tests, similarly to librbd unit tests, so potentially they may have the same problem. If you think it is still ok I am fine with it too.

Comment (wjwithagen) [reply src/test/librbd/CMakeLists.txt:161]: @tchaikov 
On FreeBSD I seem to be running OOM during linking... 
The VMs have 16Gb of memory, and stiil I manage to get OOM during linking of ceph-dencoder.
So parallelism is already set down to `-j4` to allow building.
A way to specify in cmake that some linkage is to heavy, and should be done at the end, running things serial, would be a great bonus.

Comment (tchaikov) [reply src/test/librbd/CMakeLists.txt:161]: > A way to specify in cmake that some linkage is to heavy, and should be done at the end, running things serial, would be a great bonus.

@wjwithagen not sure how to archive this. but i will see what i can do to address the linkage issue of `ceph-dencoder`. it is very different, as it includes **lots** of symbols of different structures into a monolithic executable. to address the compile time issue i split a single source file into smaller pieces in eaaec3d3193795822fb220e65a4d18b800d1b95f. but this time, we might need to go further by moving these objects into multiple shared libraries.

Comment (tchaikov) [reply src/test/librbd/CMakeLists.txt:161]: #40451

**Comment by tchaikov** [comment]
@batrick could you take a look?

**Review by batrick — APPROVED** [review]
Looks sane to me. #40451 was also an awesome change Kefu. Well done!

**Comment by tchaikov** [comment]
@batrick thanks for your review!

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/wip-cmake-job-pool/6b6f69e886e6041dec1ae001e2a7a0c1d6d999d7/

**Comment by wjwithagen** [comment]
> 
> 
> Looks sane to me. #40451 was also an awesome change Kefu. Well done!

Yes, #40451 really allows me to run much more in parallel while building.
The ninja part thus far does not reduce my compile time, but perhaps with some tweaking in this PR, I can it down even a bit more.
But the pesky crashed, which even OOMed the jenkins-server don't happen anymore.

**Pull request merged**